plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KerberizedKuduClient.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.kudu.client.KuduSession;
+import org.apache.kudu.client.KuduTable;
+import org.apache.kudu.client.ListTablesResponse;
+
+import java.io.IOException;
+
+public class KerberizedKuduClient
+        implements IKuduClient
+{
+    KuduClient client;
+    CachingKerberosAuthentication cachingKerberosAuthentication;
+
+    KerberizedKuduClient(KuduClient client, CachingKerberosAuthentication cachingKerberosAuthentication)
+    {
+        this.client = client;
+        this.cachingKerberosAuthentication = cachingKerberosAuthentication;
+    }
+
+    public KuduTable createTable(String name, Schema schema, CreateTableOptions builder)
+            throws KuduException
+    {
+        return maybeRefreshKerberosTicketAndRun(() -> client.createTable(name, schema, builder));
+    }
+
+    public DeleteTableResponse deleteTable(String name)
+            throws KuduException
+    {
+        return maybeRefreshKerberosTicketAndRun(() -> client.deleteTable(name));
+    }
+
+    public AlterTableResponse alterTable(String name, AlterTableOptions ato)
+            throws KuduException
+    {
+        return maybeRefreshKerberosTicketAndRun(() -> client.alterTable(name, ato));
+    }
+
+    public ListTablesResponse getTablesList()
+            throws KuduException
+    {
+        return maybeRefreshKerberosTicketAndRun(() -> client.getTablesList());
+    }
+
+    public ListTablesResponse getTablesList(String nameFilter)
+            throws KuduException
+    {
+        return maybeRefreshKerberosTicketAndRun(() -> client.getTablesList(nameFilter));
+    }
+
+    public boolean tableExists(String name)
+            throws KuduException
+    {
+        return maybeRefreshKerberosTicketAndRun(() -> client.tableExists(name));
+    }
+
+    public KuduTable openTable(final String name)
+            throws KuduException
+    {
+        return maybeRefreshKerberosTicketAndRun(() -> client.openTable(name));
+    }
+
+    // Fully local operation with no RPCs, no need to wrap
+    public KuduSession newSession()
+    {
+        return client.newSession();
+    }
+
+    public KuduScanner.KuduScannerBuilder newScannerBuilder(KuduTable table)
+    {
+        return client.newScannerBuilder(table);
+    }
+
+    public KuduScanToken.KuduScanTokenBuilder newScanTokenBuilder(KuduTable table)
+    {
+        return client.newScanTokenBuilder(table);
+    }
+
+    @Override
+    public void close()
+            throws KuduException
+    {
+        maybeRefreshKerberosTicketAndRun(() -> {
+            client.close();
+            return null;
+        });
+    }
+
+    @Override
+    public KuduScanner deserializeIntoScanner(byte[] serializedScanToken)
+            throws IOException
+    {
+        return maybeRefreshKerberosTicketAndRun(() ->
+                KuduScanToken.deserializeIntoScanner(serializedScanToken, client));
+    }
+
+    private <R, E extends Exception> R maybeRefreshKerberosTicketAndRun(CheckedSupplier<R, E> supplier) throws E
```

Comment:
I originally made this retry once(and only reauth if the first execution failed). However, some operations like `client.openTable()` are expected to throw a `KuduException ` if the table does not exist. Thus, retrying would cause operations to be retried needlessly.

Instead of retrying, I decided to always attempt to re-auth, this is safe as long as the operation itself does not take longer than `(1-TICKET_RENEW_WINDOW)` [20% of the ticket lifetime] which seems pretty safe for usual kerberos ticket lifetimes of ~24hours. However, this causes all client operations to be blocked on the synchronized method `reauthenticateIfSoonWillBeExpired`. The critical section is pretty small in the happy path since it just checks a timestamp, but it does introduce synchronization overhead.

What do you all think?

Reply:
does this comment still apply?

Reply:
Yes this comment still applies.

Reply:
Does KuduException have some fields or message to indicate it was a Kerberos failure instead of something else. That way you can retry only once if the failure is due to auth and lower the synchronization overhead?

Reply:
Here is that half broken commit for context showing the painpoints:
https://github.com/trinodb/trino/commit/14d757cbb66f109b9c502286c5c43e97a35e5888

We talked in person, and sounds like we are leaning towards not implementing the retrying logic and instead letting a query fail if we hit this edge case of a kerberos ticket expiring mid way through a time consuming operation.